go-controller/pkg/clustermanager/network_cluster_controller.go

```diff
@@ -208,26 +208,24 @@ func (ncc *networkClusterController) syncNodeClusterSubnet(node *corev1.Node) er
 		return err
 	}
 
-	if len(allocatedSubnets) == 0 {
+	// If allocation/validate didn't make any changes, no need to update the node
+	if len(existingSubnets) == len(hostSubnets) && len(allocatedSubnets) == 0 {
```

Comment:
@tssurya Good point, I'll add some comments.

Besides the dual->single bug, the other case this code handles is when a node's annotations have a subnet outside the ClusterNetworks or if it has a duplicate subnet already assigned to another node. Sometimes this happens when people save the node as YAML and then just `oc add` the yaml back without clearing the annotations.

In that case `[ existing > 0,  hostSubnets = 0,  allocatedSubnets > 0 ]` because AllocateNodeSubnets() will have removed the invalid/duplicate subnet and assigned a new one.

In the dual case `[ existing = 2, hostSubnets = 1, allocatedSubnets = 0 ]` because only one of the existing subnets was valid and we don't need to allocate a second one because we're now single-stack.

Reply:
> @tssurya Good point, I'll add some comments.
> 
> Besides the dual->single bug, the other case this code handles is when a node's annotations have a subnet outside the ClusterNetworks or if it has a duplicate subnet already assigned to another node. Sometimes this happens when people save the node as YAML and then just `oc add` the yaml back without clearing the annotations.
> 

ah I see! didn't think of that but sadly you are right, I seem to remember us having dealt with such a bug in the past...

> In that case `[ existing > 0, hostSubnets = 0, allocatedSubnets > 0 ]` because AllocateNodeSubnets() will have removed the invalid/duplicate subnet and assigned a new one
> 
> In the dual case `[ existing = 2, hostSubnets = 1, allocatedSubnets = 0 ]` because only one of the existing subnets was valid and we don't need to allocate a second one because we're now single-stack.

yup makes sense

Reply:
@tssurya added the comment and some testcases to cover more things.